Automattic/woocommerce-payments
# Don't load Apple/Google Pay for synced subscriptions that don't have payment due today and also require shipping

Fixes #3742

#### Changes proposed in this Pull Request










When customers attempt to use Apple Pay or Google Pay to purchase a synchronized subscription product that has no upfront costs (i.e. trial period or delayed initial payment due to product being synced) and also requires shipping, they're met with an error on the page:

![image](https://github.com/Automattic/woocommerce-payments/assets/2275145/f35a9c20-0a53-4d35-9f84-fa2f4a05760a)

Inside our `WC_Payments_Payment_Request_Button_Handler` class, we have code that prevents non-virtual (physical) products with free trials from being purchased with Apple/Google Pay, however there isn't any code to handle products that imitate a free trial period like what we do to handle synchronised subscription purchased outside of the synced date with no initial payment.

In this PR, I've introduced a new function to check if the product is a subscription product and has a free trial period (either trial set on the product or is a synced product), and requires shipping, and if so, prevent loading the Apple Pay/Google Pay buttons on the page.


#### Testing instructions





1. Enable Apple Pay and Google Pay payment methods
2. Install and active Woo Subscriptions
3. Enable synchronised subscriptions in **WooCommerce > Settings > Subscriptions > Synchronise renewals**
![image](https://github.com/Automattic/woocommerce-payments/assets/2275145/ff2a0b57-bcc5-492a-9433-952ae079a708)
4. Create a synced product that is not virtual and is not synced to today:
![image](https://github.com/Automattic/woocommerce-payments/assets/2275145/a9f7164f-c652-4acf-ae31-cf1a1b4c8ee7)
5. While on the `develop` branch, view the product page and note the Payment Request buttons (PRB) are loaded. If you attempt to purchase this product with PRB, you will see the above error.
6. Check out this branch and confirm the PRB are no longer loaded on product pages for physical synced subscription products that

Other test cases to run:
 - Physical (non-virtual) variable synced subscriptions. The PRB shouldn't show on the product page if any variation isn't supported.
 - Virtual synced subscriptions. The PRB should show.
 - Physical synced products with a sync date set to today. The PRB should show
 - Physical synced subscription that is synced to another day but proration is enabled in **WC > Settings > Subscriptions**

-------------------


**Post merge**
